### ISSUE TITLE
Error creating EXT4 filesystem

### ISSUE THREAD

**Issue by mattmoor** [diskfs/go-diskfs#274]
I am playing around with trying to construct an extremely simple and contrived disk with an EFI System Partition (FAT32) and a root partition (EXT4).

In the example below, I consistently see the following error:
```
# rm disk.raw; go run main.go 
2024/12/26 09:00:43 failed to create filesystem: wrote 0 bytes of GDT for block 0 to disk instead of expected 664403968
panic: failed to create filesystem: wrote 0 bytes of GDT for block 0 to disk instead of expected 664403968
```

If I change the root partition to FAT32 things work fine, but it's FAT32... 😉 

Here's the full example I'm playing with:
```
package main

import (
	"log"

	diskfs "github.com/diskfs/go-diskfs"
	diskpkg "github.com/diskfs/go-diskfs/disk"
	"github.com/diskfs/go-diskfs/filesystem"
	"github.com/diskfs/go-diskfs/partition/gpt"
)

func main() {
	const (
		espSize  uint64 = 100 * 1024 * 1024       // 100 MB
		diskSize uint64 = 10 * 1024 * 1024 * 1024 // 10 GB

		padding uint64 = 1024 * 1024 // 1 MB
		blkSize uint64 = uint64(diskfs.SectorSize512)

		gptGUID  = "00000000-0000-0000-0000-000000000000"
		espGUID  = "11111111-1111-1111-1111-111111111111"
		rootGUID = "22222222-2222-2222-2222-222222222222"
	)

	disk, err := diskfs.Create("disk.raw", int64(diskSize), diskfs.SectorSize512)
	if err != nil {
		log.Panicf("failed to create disk: %v", err)
	}

	if err := disk.Partition(&gpt.Table{
		GUID:          gptGUID,
		ProtectiveMBR: true,
		Partitions: []*gpt.Partition{{
			Name:  "esp",
			Type:  gpt.EFISystemPartition,
			GUID:  espGUID,
			Start: padding / blkSize,
			End:   (padding+espSize)/blkSize - 1,
			Size:  espSize, // (End - Start) * blkSize
		}, {
			Name:  "root",
			Type:  gpt.LinuxFilesystem,
			GUID:  rootGUID,
			Start: (padding + espSize) / blkSize,
			End:   (diskSize-padding)/blkSize - 1,
			Size:  diskSize - espSize - 2*padding, // (End - Start) * blkSize
		}},
	}); err != nil {
		log.Panicf("failed to partition disk: %v", err)
	}

	if _, err := disk.CreateFilesystem(diskpkg.FilesystemSpec{
		Partition:   1,
		FSType:      filesystem.TypeFat32,
		VolumeLabel: "esp",
	}); err != nil {
		log.Panicf("failed to create filesystem: %v", err)
	}

	if _, err := disk.CreateFilesystem(diskpkg.FilesystemSpec{
		Partition:   2,
		FSType:      filesystem.TypeExt4,
		VolumeLabel: "root",
	}); err != nil {
		log.Panicf("failed to create filesystem: %v", err)
	}
}
```

Note that some of the values were chosen for parity with a manually crafted example I was playing with using `fdisk` and friends.

**Comment by deitch** [comment]
Failure appears to be [here](https://github.com/diskfs/go-diskfs/blob/master/filesystem/ext4/ext4.go#L580-L586) when writing the group descriptor table for superblock backup 0 (or any other). The reason is that the data for the GDT is an empty slice, i.e. 0 bytes. 

It should be set [here[(https://github.com/diskfs/go-diskfs/blob/master/filesystem/ext4/ext4.go#L548):

```go
	g := gdt.toBytes(gdtChecksumType, sb.checksumSeed)
```

With the descriptor table set [a few lines above](https://github.com/diskfs/go-diskfs/blob/master/filesystem/ext4/ext4.go#L541):

```go
	gdt := groupDescriptors{}
```

Why is there nothing in it? I do not know. I looked at the [original commit](https://github.com/diskfs/go-diskfs/commit/f46e730062e3876265f6a7ee724b57d3202cf159#diff-2d8c12205675cead6f0a1b0fe3d76bc1e5d2dfc5908b9725e8c76971b6afc116) and it has it, so it isn't like it got removed at some point.

Is it possible the GDT never was created properly? I guess it is. I wonder if we look back on the notes we would find some "TODO" that says, "don't forget to complete creating the Group Descriptor Tables!"

Interested in trying your hand at it?

**Comment by mattmoor** [comment]
Debugging this, it seems like the `gdt` is empty here:
https://github.com/diskfs/go-diskfs/blob/31082f180e49ef0ab79d072f7fa6ae5b86009fb5/filesystem/ext4/ext4.go#L541

`toBytes` basically just walks the list of `descriptors` in the `gdt` here:
https://github.com/diskfs/go-diskfs/blob/31082f180e49ef0ab79d072f7fa6ae5b86009fb5/filesystem/ext4/groupdescriptors.go#L110-L119

However, since this is empty, `toBytes` does basically nothing?

**Comment by deitch** [comment]
Yeah, pretty much. It needs to use correct logic to fill in the desired group descriptors. Once those are there, it should do the right thing with them.

**Comment by mattmoor** [comment]
Yeah, I have no clue what goes into the descriptors, or what information they should be based on.

I am guessing that there should be `numblocks` of them, based on this:
https://github.com/diskfs/go-diskfs/blob/31082f180e49ef0ab79d072f7fa6ae5b86009fb5/filesystem/ext4/ext4.go#L554

... and that the entries will maybe be fairly formulaic based on this:
https://github.com/diskfs/go-diskfs/blob/31082f180e49ef0ab79d072f7fa6ae5b86009fb5/filesystem/ext4/ext4.go#L181-L184

The only place I see creating a `groupDescriptor` currently is here, which isn't super interecting to crib from: https://github.com/diskfs/go-diskfs/blob/31082f180e49ef0ab79d072f7fa6ae5b86009fb5/filesystem/ext4/groupdescriptors.go#L137

**Comment by deitch** [comment]
If it were a quick and easy job, I would have done it. 😃 

Check out the file ext4.md, which links to much of it

**Comment by DavidVentura** [comment]
I've just hit this as well, but there are more issues.
First, a minimal reproducer:
```
package main

import (
	"log"
	"os"

	"github.com/diskfs/go-diskfs/backend/file"
	"github.com/diskfs/go-diskfs/filesystem/ext4"
)

func main() {
	var diskSize uint64 = 10 * 1024 * 1024 * 1024 // 10 GB

	os.Truncate("disk.raw", int64(diskSize))
	f, err := os.OpenFile("disk.raw", os.O_RDWR|os.O_EXCL, 0o660)
	if err != nil {
		panic(err)
	}
	_, err = ext4.Create(file.New(f, false), int64(diskSize), 0, 512, nil)
	if err != nil {
		log.Panicf("failed to create filesystem: %v", err)
	}
}

```

I've added one "obviously broken" descriptor
```
        gdt := groupDescriptors{
                descriptors: []groupDescriptor{  
                        {unusedInodes: inodeCount - 1, size: groupDescriptorSize64Bit}, // 1 inode for `/`  
                }, 
        }  
```

but it seems like the superblocks are wrong, for this line
```
count, err = writable.WriteAt(g, incr+blockStart+int64(SuperblockSize)+start) 
```
I added logging
```
fmt.Printf("off incr %d blockstart %d sbsize %d start %d\n", incr, blockStart, int64(SuperblockSize), start)
fmt.Printf("writing at offset %d, value of size %d, want %d\n", incr+blockStart+int64(SuperblockSize)+start, len(g), gdtSize)               
```

which results in 
```
off incr 1024 blockstart 0 sbsize 1024 start 0
writing at offset 2048, value of size 64, want 671088640

```

so, for the first backup superblock, blockstart is fine (for the rest, it's absolutely not), but `gdtSize` is wrong, or at least it seems to want to write 640MB of data

**Comment by deitch** [comment]
Yeah, writing ext4 is _hard_. This one was by far the longest to implement, and still has issues.

There are links to the specs in [ext4.md](https://github.com/diskfs/go-diskfs/blob/master/filesystem/ext4/ext4.md). By all means, give it a shot. 

> 16 TiB file

No, no, not an error. I make commissions off storage charges from your cloud provider. 😆

**Comment by deitch** [comment]
@COD3HUNT3R is this specifically about ext4 and the issue listed above? If so, this is fine. If not, please open a specific separate issue with the problem.
